homeassistant/components/stream/hls.py

```diff
@@ -64,55 +78,184 @@ class HlsPlaylistView(StreamView):
     cors_allowed = True
 
     @staticmethod
-    def render_preamble(track):
-        """Render preamble."""
-        return [
-            "#EXT-X-VERSION:7",
-            f"#EXT-X-TARGETDURATION:{track.target_duration}",
-            '#EXT-X-MAP:URI="init.mp4"',
-        ]
+    def render(track, ll_hls):
+        """Render HLS playlist file."""
+        # NUM_PLAYLIST_SEGMENTS+1 because most recent is probably not yet complete
+        segments = list(track.get_segment())[-(NUM_PLAYLIST_SEGMENTS + 1) :]
 
-    @staticmethod
-    def render_playlist(track):
-        """Render playlist."""
-        segments = list(track.get_segment())[-NUM_PLAYLIST_SEGMENTS:]
+        # Create a copy of the last segment with the parts frozen for rendering playlist
+        segments[-1] = copy(segments[-1])
+        segments[-1].parts = segments[-1].parts.copy()
+
+        # Remove first segment if the last segment is actually complete
+        if segments[-1].duration:
+            segments = segments[-NUM_PLAYLIST_SEGMENTS:]
 
-        if not segments:
-            return []
+        target_duration = round(max(s.duration for s in segments if s.duration))
 
         playlist = [
+            "#EXTM3U",
+            "#EXT-X-VERSION:6",
```

Comment:
It looks like we should use version 6 instead of version 7. We use tags from 6 but not from 7.

From 6.2.1 of the RFC:

>    The value of the EXT-X-VERSION tag (Section 4.4.1.2) SHOULD NOT be
   greater than what is required for the tags and attributes in the
   Playlist (see Section 7).

From Section 7:

>    A Media Playlist MUST indicate an EXT-X-VERSION of 6 or higher if it
   contains:
>    o  The EXT-X-MAP tag in a Media Playlist that does not contain EXT-
>       X-I-FRAMES-ONLY.
> 
>    A Master Playlist MUST indicate an EXT-X-VERSION of 7 or higher if it
>    contains:
>    o  "SERVICE" values for the INSTREAM-ID attribute of the EXT-X-MEDIA
>       tag.